---
title: "mulea"
author: 
- name: "Cezary Turek"
  email: "Cezary.Turek@gmail.com"
- name: "Márton Ölbei"
- name: "Tamás Stirling"
- name: "Gergely Fekete"
- name: "Ervin Tasnádi"
- name: "Leila Gul"
- name: "Balázs Bohár"
- name: "Balázs Papp"
- name: "Wiktor Jurkowski"
- name: "Eszter Ari"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{mulea}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r global_options, include=TRUE, echo=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    warning = TRUE,
    message = TRUE,
    error = FALSE)
```

# Introduction

Traditional gene set enrichment analyses are typically limited to a few 
ontologies and do not account for the interdependence of gene sets or terms, 
resulting in overcorrected *p*-values. To address these challenges, we introduce
`mulea`, an R package offering comprehensive overrepresentation and functional 
enrichment analysis.

`mulea` employs an innovative empirical *false discovery rate* (*eFDR*) 
correction method, specifically designed for interconnected biological data, to 
accurately identify significant terms within diverse ontologies. Beyond 
conventional tools, `mulea` incorporates a wide range of ontologies encompassing
Gene Ontology, pathways, regulatory elements, genomic locations, and protein 
domains. This flexibility empowers researchers to tailor enrichment analysis to 
their specific questions, such as identifying enriched transcriptional 
regulators in gene expression data or overrepresented protein domains in protein
sets.

To facilitate seamless analysis, `mulea` provides gene sets (in standardized GMT
format) for 27 model organisms, covering 16 databases and various identifiers. 
The GMT files and the scripts we applied to create them are available at the
[GMT_files_for_mulea](https://github.com/ELTEbioinformatics/GMT_files_for_mulea)
repository. Additionally, the 
[`muleaData`](https://github.com/ELTEbioinformatics/muleaData) ExperimentData 
Bioconductor R package simplifies access to these 879 pre-defined ontologies. 
Furthermore, `mulea`'s architecture allows for easy integration of user-defined
ontologies, expanding its applicability across diverse research areas.

# Installation

After installing the `BiocManager` package, you can install 
[`fgsea`](https://bioconductor.org/packages/release/bioc/html/fgsea.html), a 
dependency for `mulea` from Bioconductor. Then, you can install `mulea` from 
this github repo using the `install_github` function of the `devtools` package:

```{r 'install', eval=FALSE, message=FALSE, warning=FALSE}
# installing the BiocManager package if needed
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
# installing the fgsea package from Bioconductor
BiocManager::install("fgsea")
# installing the mulea package from GitHub
devtools::install_github("https://github.com/ELTEbioinformatics/mulea")
```

# Example Usage

## Formatting the Differential Expression Results

This section demonstrates how to use `mulea` with a sample dataset. If you have 
your own data, feel free to skip this part and proceed directly to the 
[OverRepresentation Analysis (ORA)](#overrepresentation-analysis-ora) or 
[Gene Set Enrichment Analysis (GSEA)](#gene-set-enrichment-analysis-gsea) 
sections.

### **1. Downloading and Exploring the Dataset:**

This example analyses a differential expression (DE) dataset from a microarray 
experiment deposited in the NCBI Gene Expression Omnibus 
![GEO](../man/figures/geo_main.gif){alt="GEO" width="87"} under accession number
[GSE55662](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE55662). The 
original study by [Méhi *et al.* (2014)](https://doi.org/10.1093/molbev/msu223),
investigated the evolution of antibiotic resistance in *Escherichia coli* 
bacteria. The authors compared gene expression changes in 
*ciprofloxacin*-treated bacteria to non-treated controls.

The [GEO2R](https://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE55662) tool was used 
for differential expression analysis, comparing:

-   Non-treated control samples (2 replicates)
-   Samples treated with *ciprofloxacin* (2 replicates)

### **2. Reading the DE Results Table:**

This section would typically describe the format and key elements of the 
provided DE results table, guiding users on how to interpret the data for 
further analysis with `mulea`.

```{r 'DE1', eval=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
geo2r_result_tab <- read_tsv("GSE55662.table_wt_non_vs_cipro.tsv")
```

```{r 'DE1_eval', eval=TRUE, message=FALSE, warning=FALSE, echo=FALSE}
library(tidyverse)
geo2r_result_tab <- 
    read_tsv("../inst/extdata/GSE55662.table_wt_non_vs_cipro.tsv")
```

### **3. Examining the Data:**

We take a closer look at the first few rows of the `geo2r_result_tab` data 
frame:

```{r 'print_geo', echo=FALSE}
geo2r_result_tab %>%  
    head(3) %>% 
    knitr::kable()
```

### **4. Data Formatting:**

It's important to format the data frame appropriately for enrichment analysis. 
This process often involves steps specific to the type of microarray experiment 
conducted. In this case, we perform the following transformations:

-   **Extract Gene Symbol:** We extract the main gene symbol from the 
`Gene.symbol` column, removing any additional information that might be present.
-   **Remove Missing Values:** We remove rows where the gene symbol is missing 
(`NA`).
-   **Order by Fold Change:** We sort the data frame by the log-fold change 
(`logFC`) in descending order, prioritizing genes with the most significant 
expression changes.

```{r 'format_geo'}
geo2r_result_tab %<>% 
    # extracting the first gene symbol from the Gene.symbol column
    mutate(Gene.symbol = str_remove(
        string = Gene.symbol,
        pattern = "\\/.*")) %>% 
    # removing rows where Gene.symbol is NA
    filter(!is.na(Gene.symbol)) %>% 
    # ordering by logFC
    arrange(desc(logFC))
```

Before proceeding with enrichment analysis, we take a closer look at the first 
few rows of the formatted `geo2r_result_tab` data frame:

```{r 'print_geo_formatted', echo=FALSE}
geo2r_result_tab %>%  
    head(3) %>% 
    knitr::kable()
```

After applying these formatting steps, the data frame is ready for further 
analysis.

## The Gene Set for Identifying Regulatory Transcription Factors

This section explores the transcription factors influencing the significantly 
overexpressed genes. We employed the `mulea` package to conduct multiple 
enrichment analyses using the 
![Regulon](../man/figures/Regulon.png){alt="Regulon" width="114" height="25"} 
[database](https://regulondb.ccg.unam.mx/).

The analysis utilized a GMT file downloaded from the dedicated 
[GMT_files_for_mulea](https://github.com/ELTEbioinformatics/GMT_files_for_mulea)
GitHub repository. This file associates gene symbols with the transcription 
factors that regulate them.

The GMT file contains lists of gene symbols regulated by specific transcription 
factors, also identified by gene symbols within the file. We proceed to parse 
and analyse this data to uncover the regulatory relationships.

### 1. Downloading and Parsing the GMT File:

We can download the required GMT file from the repository using the website and 
then read the file. For this we need to call the `mulea` package.

```{r 'read_GMT1', eval=FALSE}
library(mulea)
tf_gmt <- 
    read_gmt("Transcription_factor_RegulonDB_Escherichia_coli_GeneSymbol.gmt")
```

```{r 'read_GMT2', echo=FALSE}
library(mulea)
tf_gmt <- read_gmt("../inst/extdata/Transcription_factor.gmt")
```

### 2. Let's Check Some Properties of the `tf_gmt`:

How many transcription factors it contains?

```{r 'print_nr_of_tf'}
nrow(tf_gmt)
```

The first 3 rows of the `tf_gmt`:

```{r 'print_reulon', echo=FALSE}
tf_gmt %>%
    head(3) %>% 
    knitr::kable()
```

**Important Note:** The format of this GMT differs slightly from standard GMT 
files. In the `tf_gmt`, both the `ontology_id` and `ontology_name` columns 
contain gene symbols of the transcription factors, unlike other GMT files like 
GO, where these columns hold specific identifiers and corresponding names.

Each line in the file represents a group of genes regulated by a specific 
transcription factor. The `list_of_values` column lists the gene symbols under 
the control of the transcription factor mentioned in the `ontology_id` column.

For example, to see all genes regulated by the transcription factor "AcrR", you 
can use the following code:

```{r 'print_acrr'}
tf_gmt %>% 
    # filtering the row where the ontology_id is AcrR
    filter(ontology_id == "AcrR") %>% 
    # selecting the list_of_values column
    select(list_of_values) %>% 
    # converting tibble to vector
    pull()
```

### 3. Focusing on Relevant Entries for the Enrichment Analysis

Enrichment analysis results can sometimes be skewed by overly specific or broad 
entries. `mulea` allows you to customize the size of ontology entries, ensuring 
your analysis aligns with your desired scope.

**Analysing Entry Distribution:**

Let's examine the distribution of the number of gene symbols in the 
`list_of_values` column to identify entries requiring exclusion:

```{r 'plot_list_of_values'}
Nr_of_elements_in_ontology <- tf_gmt$list_of_values %>% 
    map_dbl(length)
ggplot(mapping = aes(Nr_of_elements_in_ontology)) + 
    geom_bar() +
    theme_minimal()
```

This plot reveals entries containing over 200 gene symbols, indicating these 
transcription factors regulate too many genes, making them overly broad. We'll 
exclude them from the analysis.

Conversely, some entries hold a very small number of elements. Let's zoom in:

```{r 'plot_list_of_values_zoom', warning=FALSE}
ggplot(mapping = aes(Nr_of_elements_in_ontology)) + 
    geom_bar() +
    xlim(0, 15) +
    theme_minimal()
```

**Filtering Entries:**

Based on our observations, we'll exclude entries with less than 3 or more than 
400 gene symbols. Let's check the remaining number of transcription factors:

```{r 'exclude_ontology'}
tf_gmt_filtered <- filter_ontology(
    gmt = tf_gmt,
    min_nr_of_elements = 3,
    max_nr_of_elements = 400)
```

**Results:**

We can now determine the number of transcription factors remaining in the 
filtered dataset:

```{r 'print_nr_of_tf_filtered'}
nrow(tf_gmt_filtered)
```

It is possible to write the filtered ontology as a GMT file using the 
`write_gmt` function.

```{r 'save_gmt', eval=FALSE}
write_gmt(gmt = tf_gmt_filtered, file = "Filtered.gmt")
```

### Converting a List to an Ontology Object

The `mulea` package provides a function to convert a list of gene sets to an 
ontology (GMT) object. This function is called `list_to_gmt`. The following 
example demonstrates how to use this function:

```{r 'list_to_gmt_example', eval=FALSE}
# creating a list of gene sets
ontology_list <- list(
    gene_set1 = c("gene1", "gene2", "gene3"),
    gene_set2 = c("gene4", "gene5", "gene6"))

# converting the list to a ontology (GMT) object
new_ontology_object <- list_to_gmt(ontology_list)
```

## OverRepresentation Analysis (ORA) {#overrepresentation-analysis-ora}

This approach analyses groups of genes (sets) to identify if they are enriched 
in specific categories – transcription factors – within a reference set. It 
requires two key elements:

1.  **Target set:** This contains genes of interest, such as significantly 
overexpressed genes in our experiment.

2.  **Background set:** This represents the broader context, often including all
genes investigated in our study.

To ensure meaningful results, a clear threshold needs to be applied beforehand. 
This could involve filtering genes based on corrected *p*-values, *z*-scores 
(commonly set at 0.05), or fold-change values (*e.g.*, a minimum 2-fold change).

`mulea` utilizes the hypergeometric test to assess overrepresentation within 
categories. This test is similar to the lower-tailed Fisher's exact test and 
helps determine if the observed enrichment is statistically significant by 
considering both the target and background sets.

### Preparing the Target set

A vector containing the gene symbols of significantly overexpressed (adjusted 
*p*-value \< 0.05) genes with greater than 2 fold-change (logFC \> 1).

```{r 'target_set'}
sign_genes <- geo2r_result_tab %>% 
    # filtering for adjusted p-value < 0.05 and logFC > 1
    filter(adj.P.Val < 0.05 & logFC > 1) %>% 
    # selecting the Gene.symbol column
    select(Gene.symbol) %>% 
    # converting the tibble to a vector
    pull() %>% 
    # removing duplicates
    unique()
```

The first 10 elements of the target set:

```{r 'target_head'}
sign_genes %>% 
    head(10)
```

The number of genes in the target set:

```{r 'target_gene_nr'}
sign_genes %>% 
    length()
```

### Preparing the Background Set

A vector containing the gene symbols of all genes were included in the 
differential expression analysis.

```{r 'background_set'}
background_genes <- geo2r_result_tab %>% 
    # selecting the Gene.symbol column
    select(Gene.symbol) %>% 
    # convertin the tibble to a vector
    pull() %>% 
    # removing duplicates
    unique()
```

The number of genes in the background set:

```{r 'background_gene_nr'}
background_genes %>% 
    length()
```

### Performing the OverRepresentation Analysis

To perform the analysis, we will first establish a model using the `ora` 
function. This model defines the parameters for the enrichment analysis. 
Subsequently, we will execute the test itself using the `run_test` function. It 
is important to note that for this example, we will employ 10,000 permutations 
for the *empirical false discovery rate* correction, which is the recommended 
minimum, to ensure robust correction for multiple testing.

```{r 'ora'}
# creating the ORA model using the GMT variable
ora_model <- ora(
    gmt = tf_gmt_filtered, 
    # the test set variable
    element_names = sign_genes,
    # the background set variable
    background_element_names = background_genes,
    # the p-value adjustment method
    p_value_adjustment_method = "eFDR",
    number_of_permutations = 10000,
    # the number of processor threads to use
    nthreads = 2) 
# running the ORA
ora_results <- run_test(ora_model)
```

### Examining the ORA Result

The `ora_results` data frame summarizes the enrichment analysis, listing 
enriched ontology entries – in our case transcription factors – alongside their
associated *p*-values and *empirical FDR* values. We can now determine the 
number of transcription factors classified as "enriched" based on these 
statistical measures (*eFDR* \< 0.05).

```{r 'ora_size'}
ora_results %>%
    # rows where the eFDR < 0.05
    filter(eFDR < 0.05) %>% 
    # the number of such rows
    nrow()
```

And inspect the significant results:

```{r 'print_ora', eval=FALSE}
ora_results %>%
    # arrange the rows by the eFDR values
    arrange(eFDR) %>% 
    # rows where the eFDR < 0.05
    filter(eFDR < 0.05)
```

```{r 'print_ora2', echo=FALSE}
ora_results %>%
    # arrange the rows by the eFDR values
    arrange(eFDR) %>% 
    # rows where the eFDR < 0.05
    filter(eFDR < 0.05) %>% 
    knitr::kable()
```

### Visualizing the ORA Result

For a more comprehensive understanding of the enriched transcription factors, 
`mulea` provides diverse visualization tools, including lollipop charts, 
networks, and heatmaps. These visualizations can effectively reveal patterns and
relationships among the enriched factors.

Initializing the visualization with the `reshape_results` function:

```{r 'init_plot_ora'}
ora_reshaped_results <- reshape_results(
    model = ora_model,
    model_results = ora_results,
    # choosing which column to use for the indication of significance
    p_value_type_colname = "eFDR")
```

**Visualizing the spread of *eFDR* values: Lollipop plot**

Lollipop charts offer a graphical representation of the distribution of enriched
transcription factors. The *y*-axis displays the transcription factors, while 
the *x*-axis represents their corresponding *eFDR* values. The dots are is 
coloured based on their significance level. This visualization helps us examine
the spread of *eFDR*s and identify factors exceeding the commonly used 
significance threshold of 0.05.

```{r 'lollipop_plot_ora'}
plot_lollipop(
    reshaped_results = ora_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "eFDR")
```

**Visualizing the spread of *eFDR* values: Bar plot**

Bar charts offer very similar graphical representation of the distribution of 
enriched transcription factors as the lollipop plot. The *y*-axis displays the 
transcription factors, while the *x*-axis represents their corresponding *eFDR*
values. The bars are is coloured based on their significance level. This 
visualization helps us examine the spread of *eFDR*s and identify factors 
exceeding the commonly used significance threshold of 0.05.

```{r 'bar_plot_ora'}
plot_barplot(
    reshaped_results = ora_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "eFDR")
```

**Visualizing Relationships: Network Plot**

This function generates a network visualization of the enriched transcription 
factors. Each node represents a transcription factor and is coloured based on 
its significance level. A connection (edge) is drawn between two nodes if they 
share at least one common gene belonging to the **target set**, meaning that 
both transcription factors regulate the expression of the same target gene. The
thickness of the edge reflects the number of shared genes belonging to the 
**target set**.

```{r 'network_plot_ora'}
plot_graph(
    reshaped_results = ora_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "eFDR")
```

**Heatmap**

The heatmap displays the genes associated with the enriched transcription 
factors. Each row represents a transcription factor and is coloured based on its
significance level. Each column represents a target gene belonging to the 
**target set** that is potentially regulated by one or more of the enriched 
transcription factors.

```{r 'heatmap_ora'}
plot_heatmap(
    reshaped_results = ora_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # column that indicates the significance values
    p_value_type_colname = "eFDR")
```

## Gene Set Enrichment Analysis (GSEA) {#gene-set-enrichment-analysis-gsea}

To perform enrichment analysis using ranked lists, you need to provide an 
ordered list of elements, such as genes, transcripts, or proteins. This ranking
is typically determined by the results of your prior analysis, potentially based
on factors like *p*-values, *z*-scores, fold-changes, or others. Crucially, the
ranked list should include all elements involved in your analysis. For instance,
in a differential expression study, it should encompass all genes that were 
measured.

`mulea` utilizes the Kolmogorov-Smirnov approach with a permutation test to 
calculate gene set enrichment analyses. This functionality is implemented 
through the integration of the 
[`fgsea`](https://bioconductor.org/packages/release/bioc/html/fgsea.html) 
Bioconductor R package.

GSEA requires input data about the genes analysed in our experiment. This data 
can be formatted in two ways:

1.  **Data frame:** This format should include all genes investigated and their
respective log fold change values (or other values for ordering the genes) 
obtained from the differential expression analysis.

2.  **Two vectors:** Alternatively, you can provide two separate vectors. One 
vector should contain the gene symbols (or IDs), and the other should hold the 
corresponding log fold change values (or other values for ordering the genes) 
for each gene.

### Preparing the data frame input for the GSEA

```{r 'gsea_input'}
# if there are duplicated Gene.symbols keep the first one only
geo2r_result_tab_filtered <- geo2r_result_tab %>% 
    # grouping by Gene.symbol to be able to filter
    group_by(Gene.symbol) %>%
    # keeping the first row for each Gene.symbol from rows with the same 
    # Gene.symbol
    filter(row_number()==1) %>% 
    # ungrouping
    ungroup() %>% 
    # arranging by logFC in descending order
    arrange(desc(logFC)) %>%
    select(Gene.symbol, logFC)
```

The number of gene symbols in the `geo2r_result_tab_filtered` vector:

```{r 'ordered_genes_length'}
geo2r_result_tab_filtered %>% 
    nrow()
```

### Performing the Gene Set Enrichment Analysis

To perform the analysis, we will first establish a model using the `gsea` 
function. This model defines the parameters for the enrichment analysis. 
Subsequently, we will execute the test itself using the `run_test` function. We
will employ 10,000 permutations for the false discovery rate correction, to 
ensure robust correction for multiple testing.

```{r 'gsea', warning=FALSE, message=FALSE}
# creating the GSEA model using the GMT variable
gsea_model <- gsea(
    gmt = tf_gmt_filtered,
    # the names of elements to test
    element_names = geo2r_result_tab_filtered$Gene.symbol,
    # the logFC-s of elements to test
    element_scores = geo2r_result_tab_filtered$logFC,
    # consider elements having positive logFC values only
    element_score_type = "pos",
    # the number of permutations
    number_of_permutations = 10000)
# running the GSEA
gsea_results <- run_test(gsea_model)
```

### Examining the GSEA Results

The `gsea_results` data frame summarizes the enrichment analysis, listing 
enriched ontology entries – in our case transcription factors – alongside their
associated *p*-values and adjusted *p*-value values. We can now determine the 
number of transcription factors classified as "enriched" based on these 
statistical measures (adjusted *p*-value \< 0.05).

```{r 'gsea_size'}
gsea_results %>%
    # rows where the adjusted_p_value < 0.05
    filter(adjusted_p_value < 0.05) %>% 
    # the number of such rows
    nrow()
```

And inspect the significant results:

```{r 'print_gsea', eval=FALSE}
gsea_results %>%
    # arrange the rows by the adjusted_p_value values
    arrange(adjusted_p_value) %>% 
    # rows where the adjusted_p_value < 0.05
    filter(adjusted_p_value < 0.05)
```

```{r 'print_gsea2', echo=FALSE}
gsea_results %>%
    # arrange the rows by the adjusted_p_value values
    arrange(adjusted_p_value) %>% 
    # rows where the adjusted_p_value < 0.05
    filter(adjusted_p_value < 0.05) %>% 
    knitr::kable()
```

### Visualizing the GSEA Results

Initializing the visualization with the `reshape_results` function:

```{r 'init_plot_gsea'}
gsea_reshaped_results <- reshape_results(
    model = gsea_model,
    model_results = gsea_results,
    # choosing which column to use for the indication of significance
    p_value_type_colname = "adjusted_p_value")
```

**Visualizing the spread of adjusted *p*-values: Lollipop plot**

Lollipop charts offer a graphical representation of the distribution of enriched
transcription factors. The *y*-axis displays the transcription factors, while 
the *x*-axis represents their corresponding adjusted *p*-values. The dots are is
coloured based on their significance level. This visualization helps us examine 
the spread of adjusted *p*-values and identify factors exceeding the commonly 
used significance threshold of 0.05.

```{r 'lollipop_plot_gsea'}
plot_lollipop(
    reshaped_results = gsea_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "adjusted_p_value")
```

**Visualizing the spread of adjusted *p*-values: Bar plot**

Bar charts offer very similar graphical representation of the distribution of 
enriched transcription factors as the lollipop plot. The *y*-axis displays the
transcription factors, while the *x*-axis represents their corresponding 
adjusted *p*-values. The bars are is coloured based on their significance 
level. This visualization helps us examine the spread of adjusted *p*-values and
identify factors exceeding the commonly used significance threshold of 0.05.

```{r 'bar_plot_gsea'}
plot_barplot(
    reshaped_results = gsea_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "adjusted_p_value")
```

**Visualizing Relationships: Network Plot**

This function generates a network visualization of the enriched transcription 
factors. Each node represents a transcription factor and is coloured based on 
its significance level. A connection (edge) is drawn between two nodes if they 
share at least one common gene belonging to the **ranked list**, meaning that 
both transcription factors regulate the expression of the same target gene. The
thickness of the edge reflects the number of shared genes belonging to the 
**ranked list**.

```{r 'network_plot_gsea'}
plot_graph(
    reshaped_results = gsea_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # upper threshold for the value indicating the significance
    p_value_max_threshold = 0.05,
    # column that indicates the significance values
    p_value_type_colname = "adjusted_p_value")
```

**Heatmap**

The heatmap displays the genes associated with the enriched transcription 
factors. Each row represents a transcription factor and is coloured based on its
significance level. Each column represents a target gene belonging to the 
**ranked list** that is potentially regulated by one or more of the enriched 
transcription factors. There are too many genes belonging to each transcription
factor, therefore heatmap visualization is less optimal in this case.

```{r 'heatmap_gsea', fig.width=20}
plot_heatmap(
    reshaped_results = gsea_reshaped_results,
    # the column containing the names we wish to plot
    ontology_id_colname = "ontology_id",
    # column that indicates the significance values
    p_value_type_colname = "adjusted_p_value")
```

# Session Info

```{r 'session_info'}
sessionInfo()
```

# How to Cite the `mulea` Package?

To cite package `mulea` in publications use:

Turek C, Ölbei M, Stirling T, Fekete G, Tasnádi E, Gul L, Bohár B, Papp B, 
Jurkowski W, Ari E (2024) mulea - an R package for enrichment analysis using 
various ontologies and empirical FDR correction. R package version 0.99.0, 
<https://github.com/ELTEbioinformatics/mulea>.
